README.md
# CFOF (Concentration Free Outlier Factor)

🚧 **Work In Progress..**

Python implementation of Concentration Free Outlier Factor (CFOF) [[1]](#1).

## CFOF properties

- [**Concentration**](## "The tendency of distances to become almost indiscernible as dimensionality
increases.") free
- Does not suffer of the **hubness** problem
- [**Semi–locality**](## "CFOF score is both translation and scale-invariant and, hence, that the number of outliers coming from each cluster is directly proportional to its size and to its kurtosis")
- **fast-CFOF** algorithm allows to calculate reliably CFOF scores with linear cost both in the dataset size and dimensionality

## Installation

To install the latest release:
```
$ pip install cfof
```

## Usage

Import `CFOF` and `FastCFOF`.

```python
>>> from cfof import CFOF, FastCFOF
>>> import numpy as np
```

Load data.

```python
>>> X = np.array([[-1, -1], [-2, -1], [-3, -2], [1, 1], [2, 1], [3, 2]])
```

Instantiate `CFOF` or `FastCFOF`, then call `.compute(X)` to calculate the scores. `.compute(X)` returns `sc`, where `sc[i, l]` is score of object `i` for `ϱ_l` (rhos[l]).

You can also calculate CFOF scores from a precomputed distance matrix using 
`.compute_from_distance_matrix()`.

### CFOF (hard-CFOF)

Use `compute` to compute CFOF scores directly from data.

```python
>>> cfof_clf = CFOF(metric='euclidean', rhos=[0.5, 0.6], n_jobs=1)
>>> cfof_clf.compute(X)
array([[0.5       , 0.66666667],
       [0.33333333, 0.83333333],
       [0.5       , 1.        ],
       [0.5       , 0.66666667],
       [0.33333333, 0.83333333],
       [0.5       , 1.        ]])
```

Use `compute_from_distance_matrix` to compute CFOF scores from a precomputed 
distance matrix.

```python
>>> from sklearn.metrics import pairwise_distances
>>> distance_matrix = pairwise_distances(X, metric='euclidean')
>>> cfof_clf.compute_from_distance_matrix(distance_matrix)
array([[0.5       , 0.66666667],
       [0.33333333, 0.83333333],
       [0.5       , 1.        ],
       [0.5       , 0.66666667],
       [0.33333333, 0.83333333],
       [0.5       , 1.        ]])
```

### FastCFOF (soft-CFOF)

Use `compute` to compute CFOF scores directly from data.

```python
>>> np.random.seed(10)
>>> X = np.random.randint(0, 100, size=(1000, 3))
>>>
>>> fast_cfof_clf = FastCFOF(metric='euclidean',
...                          rhos=[0.001, 0.005, 0.01, 0.05, 0.1],
...                          epsilon=0.1, delta=0.1, n_bins=50, n_jobs=1)
>>> fast_cfof_clf.compute(X)
array([[0.00954095, 0.00954095, 0.01930698, 0.05963623, 0.10481131],
       [0.00954095, 0.00954095, 0.01930698, 0.06866488, 0.10481131],
       [0.00954095, 0.00954095, 0.02559548, 0.06866488, 0.10481131],
       ...,
       [0.00954095, 0.00954095, 0.01930698, 0.05963623, 0.10481131],
       [0.00954095, 0.00954095, 0.03393222, 0.15998587, 0.24420531],
       [0.00954095, 0.00954095, 0.02559548, 0.0390694 , 0.09102982]])
```

Use `compute_from_distance_matrix` to compute CFOF scores from a precomputed 
distance matrix.

```python
>>> from sklearn.metrics import pairwise_distances
>>> distance_matrix = pairwise_distances(X, metric='euclidean')
>>> fast_cfof_clf.compute_from_distance_matrix(distance_matrix)
array([[0.00954095, 0.00954095, 0.01930698, 0.05963623, 0.10481131],
       [0.00954095, 0.00954095, 0.01930698, 0.06866488, 0.10481131],
       [0.00954095, 0.00954095, 0.02559548, 0.06866488, 0.10481131],
       ...,
       [0.00954095, 0.00954095, 0.01930698, 0.05963623, 0.10481131],
       [0.00954095, 0.00954095, 0.03393222, 0.15998587, 0.24420531],
       [0.00954095, 0.00954095, 0.02559548, 0.0390694 , 0.09102982]])
```

### CFOFiSAX

This library provides a wrapper for [pyCFOFiSAX](https://github.com/luk-f/pyCFOFiSAX) [[2]](#2)

```python
>>> from cfof.cfof_isax import CFOFiSAXWrapper
```

Refer to [`pyCFOFiSAX` documentation](https://pycfofisax.readthedocs.io/fr/main/) 
for more details.

### TODOs

- [ ] Add support for [`faiss`](https://github.com/facebookresearch/faiss) (GPU).
- [ ] Parallelize FastCFOF.
- [ ] Add unit tests.
- [ ] Add benchmarks.
- [ ] Wrap [pyCFOFiSAX](https://github.com/luk-f/pyCFOFiSAX).

## References

<a id="1">[1]</a>
ANGIULLI, Fabrizio. CFOF: a concentration free measure for anomaly detection. ACM Transactions on Knowledge Discovery from Data (TKDD), 2020, vol. 14, no 1, p. 1-53.

<a id="2">[2]</a>
FOULON, Lucas, FENET, Serge, RIGOTTI, Christophe, et al. Scoring Message Stream Anomalies in Railway Communication Systems. In : 2019 International Conference on Data Mining Workshops (ICDMW). IEEE, 2019. p. 769-776.


CHANGELOG.md
## Version 0.4.0 (2021-10-22)
- Add some basic unit tests.
- Add `compute_from_distance_matrix` for CFOF and FastCFOF.
- Add some `mypy` fixes.
- Add `CFOFiSAX` wrapper.

## Version 0.3.0 (2021-10-15)
- FastCFOF
  - Numpyfy `_fast_cfof_part`

## Version 0.2.0 (2021-10-15)
- FastCFOF
  - Fix `_k_bin_inv`.
  - Optimize `_k_bin`.
  - Expose `c` argument.
  - Add `partition_size` argument.

## Version 0.1.0 (2021-10-11)
- 🎉 first release!


tests/test_cfof.py
import numpy as np
from sklearn.metrics import pairwise_distances

from cfof import CFOF


class TestCFOF:
    def test_cfof_simple(self):
        expected_scores = np.array([[0.5,
                                     0.66666667], [0.33333333, 0.83333333],
                                    [0.5, 1.], [0.5, 0.66666667],
                                    [0.33333333, 0.83333333], [0.5, 1.]])

        X = np.array([[-1, -1], [-2, -1], [-3, -2], [1, 1], [2, 1], [3, 2]])
        cfof_clf = CFOF(metric='euclidean', rhos=[0.5, 0.6], n_jobs=1)
        scores = cfof_clf.compute(X)

        assert np.allclose(expected_scores, scores)

    def test_cfof_compute_and_compute_from_same_result(self):
        X = np.array([[-1, -1], [-2, -1], [-3, -2], [1, 1], [2, 1], [3, 2]])

        cfof_clf = CFOF(metric='euclidean', rhos=[0.5, 0.6], n_jobs=1)

        scores = cfof_clf.compute(X)

        distance_matrix = pairwise_distances(X, metric='euclidean')
        scores_from_distance_matrix = cfof_clf.compute_from_distance_matrix(
            distance_matrix)

        assert np.allclose(scores_from_distance_matrix, scores)


tests/test_fast_cfof.py
import numpy as np
from sklearn.metrics import pairwise_distances

from cfof import FastCFOF


class TestFastCFOF:
    def test_fast_cfof_simple(self):
        expected_score_0_0 = 0.00954095

        np.random.seed(10)
        X = np.random.randint(0, 100, size=(1000, 3))

        fast_cfof_clf = FastCFOF(metric='euclidean',
                                 rhos=[0.001, 0.005, 0.01, 0.05, 0.1],
                                 epsilon=0.1,
                                 delta=0.1,
                                 n_bins=50,
                                 n_jobs=1)

        scores = fast_cfof_clf.compute(X)

        assert np.isclose(scores[0][0], expected_score_0_0)

    def test_fast_cfof_compute_and_compute_from_same_result(self):
        np.random.seed(10)
        X = np.random.randint(0, 100, size=(1000, 3))

        fast_cfof_clf = FastCFOF(metric='euclidean',
                                 rhos=[0.001, 0.005, 0.01, 0.05, 0.1],
                                 epsilon=0.1,
                                 delta=0.1,
                                 n_bins=50,
                                 n_jobs=1)

        scores = fast_cfof_clf.compute(X)

        distance_matrix = pairwise_distances(X, metric='euclidean')
        scores_from_distance_matrix = fast_cfof_clf.compute_from_distance_matrix(
            distance_matrix)

        assert np.allclose(scores_from_distance_matrix, scores)


src/cfof/cfof.py
from multiprocessing import Pool
from typing import Callable, List, Union

import numpy as np
from sklearn.neighbors import NearestNeighbors


class CFOF:
    """
    Concentration Free Outlier Factor.

    Parameters
    ----------
    metric : str or callable, default 'euclidean'
        Must be a valid `sklearn.metrics`.
    rhos : List[float], default [0.01]
        `ϱ` parameters, fraction of the data population.
        Must be between 0 and 1.
    n_jobs : int, default None
        The number of jobs to use for the computation.
        `-1` means using all processors.

    References
    ----------
    .. [1] Angiulli, F. (2020, January).
           CFOF: A Concentration Free Measure for Anomaly Detection.
           In ACM Transactions on Knowledge Discovery from Data.
    """
    def __init__(self,
                 metric: Union[str, Callable] = 'euclidean',
                 rhos: List[float] = [0.01],
                 n_jobs=None) -> None:
        self.rhos = np.array(rhos)
        if (self.rhos < 0.0).any() or (self.rhos > 1.0).any():
            raise ValueError(f'rhos ({self.rhos}) must be between 0 and 1')

        self.metric = metric
        self.n_jobs = n_jobs

        self.thresholds = None
        self.n = None

    def compute(self, X: np.ndarray) -> np.ndarray:
        """
        Compute hard-CFOF scores.

        Parameters
        ----------
        X : numpy.ndarray
            Dataset.

        Returns
        -------
        numpy.ndarray
            CFOF scores `sc`.
            sc[i, l] is score of object `i` for `ϱl` (rhos[l]).
        """
        self._update_params(X)
        neighbors = self._find_neighbors(X)
        return self._cfof(neighbors)

    def compute_from_distance_matrix(
            self, distance_matrix: np.ndarray) -> np.ndarray:
        """
        Compute hard-CFOF scores from distance matrix.

        Parameters
        ----------
        compute_from_distance_matrix : numpy.ndarray
            Distance matrix.

        Returns
        -------
        numpy.ndarray
            CFOF scores `sc`.
            sc[i, l] is score of object `i` for `ϱl` (rhos[l]).
        """
        self._update_params(distance_matrix)
        neighbors = self._find_neighbors_from_distance_matrix(distance_matrix)
        return self._cfof(neighbors)

    def _update_params(self, input: np.ndarray):
        self.n, _ = input.shape
        self.thresholds = self.rhos * self.n

    # `neighbors[i]` are the neighbours of i in order of proximity.
    # Every point is the first neighbor of itself.
    def _cfof(self, neighbors: np.ndarray) -> np.ndarray:
        # `min_k_neighborhood[i, j]` represents min k such that i contains j
        # in its neighborhood.
        min_k_neighborhood = np.argsort(neighbors) + 1

        with Pool(processes=self.n_jobs) as pool:
            sc = pool.map(self._compute_col_cfof, min_k_neighborhood.T)

        sc = np.array(sc) / self.n

        return sc

    def _find_neighbors(self, X: np.ndarray, algorithm='auto'):
        nbrs = NearestNeighbors(n_neighbors=len(X),
                                algorithm=algorithm,
                                metric=self.metric,
                                n_jobs=self.n_jobs).fit(X)
        indices = nbrs.kneighbors(X, return_distance=False)
        return indices

    def _find_neighbors_from_distance_matrix(self,
                                             distance_matrix: np.ndarray):
        indices = np.argsort(distance_matrix)
        return indices

    def _compute_col_cfof(self, col):
        counter = np.bincount(col).cumsum()
        return np.array([np.argmax(counter >= t) for t in self.thresholds])


src/cfof/cfof_isax.py
from typing import List

import numpy as np
from pyCFOFiSAX import CFOFiSAX


# TODO: improve and check this.
class CFOFiSAXWrapper:
    """
    CFOFiSAX wrapper.

    Parameters
    ----------
    reference_data : numpy.array
        Refrence data.
    rhos : List[float], default [0.01]
        `ϱ` parameters, fraction of the data population.
        Must be between 0 and 1.
    verbose : bool, default False
        Verbosity.

    Notes
    -----
        Refer to `pyCFOFiSAX` 
        `documentation <https://pycfofisax.readthedocs.io/fr/main/>`_ 
        for more details.

    References
    ----------
    .. [2] FOULON, Lucas, FENET, Serge, RIGOTTI, Christophe, et al. 
           Scoring Message Stream Anomalies in Railway Communication Systems. 
           In : 2019 International Conference on Data Mining Workshops (ICDMW). 
           IEEE, 2019. p. 769-776.
    """
    def __init__(self,
                 reference_data: np.array,
                 rhos: List[float] = [0.001, 0.005, 0.01, 0.05, 0.1],
                 size_word: int = 200,
                 threshold: int = 30,
                 base_cardinality: int = 2,
                 number_tree: int = 20,
                 verbose=False) -> None:

        self.reference_data = reference_data
        self.rhos = rhos

        self.cfof_isax = CFOFiSAX()
        self.cfof_isax.init_forest_isax(data_ts=self.reference_data,
                                        size_word=size_word,
                                        threshold=threshold,
                                        base_cardinality=base_cardinality,
                                        number_tree=number_tree)
        self.cfof_isax.forest_isax.index_data(self.reference_data)
        self.cfof_isax.forest_isax.preprocessing_forest_for_icfof(
            self.reference_data, bool_print=verbose, count_num_node=verbose)

    def compute(self, X: np.ndarray) -> np.ndarray:
        """
        Compute iSAX CFOF scores.

        Parameters
        ----------
        X : numpy.ndarray
            Dataset.

        Returns
        -------
        numpy.ndarray
            CFOF scores `sc`.
            sc[i, l] is score of object `i` for `ϱl` (rhos[l]).
        """
        sc = self.cfof_isax.score_icfof(X,
                                        self.reference_data,
                                        rho=self.rhos,
                                        each_tree_score=True,
                                        fast_method=True)
        return sc


src/cfof/fast_cfof.py
from typing import Callable, List, Optional, Union

import numpy as np
from sklearn.metrics import pairwise_distances


class FastCFOF:
    """
    Fast-CFOF.

    Fast-CFOF exploits sampling to avoid the computation of exact nearest
    neighbors. The cost of fast-CFOF is linear both in the dataset size and
    dimensionality.

    Parameters
    ----------
    metric : str or callable, default 'euclidean'
        Must be a valid `sklearn.metrics`.
    rhos : List[float], default [0.001, 0.005, 0.01, 0.05, 0.1]
        `ϱ` parameters, fraction of the data population.
        Must be between 0 and 1.
    epsilon : float, default 0.01
        ϵ, absolute error. (0 < ϵ < 1)
    delta : float, default 0.01
        δ, error probability. (0 < δ < 1)
    c : int, default 1
        c, with c ∈ [0, 3].
    n_bins : int, default 10
        Histogram bins.
    partition_size : Optional[int], default None
        Size of partitions.
        If None, it is computed using ϵ and δ.
        If given, ϵ and δ are ignored.
    n_jobs : int, default None
        The number of jobs to use for the computation.
        `-1` means using all processors.

    References
    ----------
    .. [1] Angiulli, F. (2020, January).
           CFOF: A Concentration Free Measure for Anomaly Detection.
           In ACM Transactions on Knowledge Discovery from Data.
    """
    def __init__(self,
                 metric: Union[str, Callable] = 'euclidean',
                 rhos: List[float] = [0.001, 0.005, 0.01, 0.05, 0.1],
                 epsilon: float = 0.01,
                 delta: float = 0.01,
                 c: int = 1,
                 n_bins: int = 10,
                 partition_size: Optional[int] = None,
                 n_jobs=None) -> None:
        self.metric = metric

        self.rhos = np.array(rhos)
        if (self.rhos < 0.0).any() or (self.rhos > 1.0).any():
            raise ValueError(f'rhos ({self.rhos}) must be between 0 and 1')

        if not (0 < epsilon < 1):
            raise ValueError(f'epsilon ({epsilon}) must be between 0 and 1')

        if not (0 < delta < 1):
            raise ValueError(f'delta ({delta}) must be between 0 and 1')

        if not (0 <= c <= 3):
            raise ValueError(f'c ({c}) must be between 0 and 3')
        self.c = c

        if not n_bins > 0:
            raise ValueError(f'n_bins ({n_bins}) must be positive')
        self.n_bins = n_bins

        if partition_size is None:
            self.s = int(np.ceil((1 / (2 * (epsilon**2))) * np.log(2 / delta)))
        else:
            self.s = partition_size

        self.n_jobs = n_jobs

        self.binning_ratio = None
        self.binning_ratio_log = None
        self.n = None

        self.X = None
        self.distance_matrix = None

        # sc[i, l] is score of object `i` for `ϱl` (rhos[l]).
        self.sc = None

    def compute(self, X: np.ndarray) -> np.ndarray:
        """
        Compute soft-CFOF scores.

        Parameters
        ----------
        X : numpy.ndarray
            Dataset.

        Returns
        -------
        numpy.ndarray
            CFOF scores `sc`.
            sc[i, l] is score of object `i` for `ϱl` (rhos[l]).
        """
        self.X = X
        self._fast_cfof()
        self.X = None
        return self.sc

    def compute_from_distance_matrix(
            self, distance_matrix: np.ndarray) -> np.ndarray:
        """
        Compute hard-CFOF scores from distance matrix.

        Parameters
        ----------
        distance_matrix : numpy.ndarray
            Distance matrix.

        Returns
        -------
        numpy.ndarray
            CFOF scores `sc`.
            sc[i, l] is score of object `i` for `ϱl` (rhos[l]).
        """
        self.distance_matrix = distance_matrix
        self._fast_cfof()
        self.distance_matrix = None
        return self.sc

    def _update_params(self):
        if self.X is not None:
            self.n, _ = self.X.shape
        elif self.distance_matrix is not None:
            self.n, _ = self.distance_matrix.shape

        self.binning_ratio = self.n**(1 / (self.n_bins - 1))
        self.binning_ratio_log = np.log(self.binning_ratio)
        self.sc = np.zeros((self.n, len(self.rhos)))

    def _fast_cfof(self):
        self._update_params()

        i = 0

        if self.s > self.n:
            raise ValueError(
                f"Partition (s = {self.s}) can't be bigger than dataset (n = {self.n})"
            )

        while i < self.n:
            if i + self.s < self.n:
                a = i
            else:
                a = self.n - self.s
            b = a + self.s
            self._fast_cfof_part(start_i=a, end_i=b)
            i = i + self.s

    def _fast_cfof_part(self, start_i: int, end_i: int):
        if self.X is not None:
            # Distances computation
            distances = pairwise_distances(self.X[start_i:end_i],
                                           n_jobs=self.n_jobs)
        elif self.distance_matrix is not None:
            distances = self.distance_matrix[start_i:end_i, start_i:end_i]

        s = end_i - start_i

        hst = np.zeros((s, self.n_bins))

        # Nearest neighbor count estimation
        for i in range(s):
            dst = distances[i, :]

            # Count update
            ord = np.argsort(dst)

            p_s = (np.arange(s) + 1) / s
            k_up_s = np.floor(self.n * p_s + self.c * np.sqrt(self.n * p_s *
                                                              (1 - p_s)) + 0.5)
            k_pos_s = self._k_bin(k_up_s)
            hst[ord, k_pos_s] += 1

        # Scores computation
        for i in range(s):
            count = 0
            k_pos = 0

            for l, rho in enumerate(self.rhos):
                while count < s * rho:
                    count += hst[i, k_pos]
                    k_pos += 1

                self.sc[start_i + i, l] = self._k_bin_inv(k_pos) / self.n

    def _k_bin(self, k_up):
        return (np.log(k_up) / self.binning_ratio_log).astype(int)

    def _k_bin_inv(self, k_pos):
        return self.binning_ratio**k_pos
